### ISSUE TITLE
Gromacs topology files do not keep correct atom type names

### ISSUE THREAD

**Issue by tlfobe** [openforcefield/openff-interchange#767]
**Description**
I've run into a weird error with Interchange where when I write a Gromacs topology file from an Interchange object atom type names are not preserved in the final `.top` file. This can be seen in this short block of the `[ atomtypes ]` block below:

```
[ atomtypes ]
;type, bondingtype, atomic_number, mass, charge, ptype, sigma, epsilon
mom_octamer_mbuild.pdb_0	     6	12.01078	0.0000000000000000	A    	0.337953176162662	0.4553891161106184
mom_octamer_mbuild.pdb_1	     8	15.99943	0.0000000000000000	A    	0.3025106490435313	0.7048581468486769
mom_octamer_mbuild.pdb_2	     6	12.01078	0.0000000000000000	A    	0.3480646886945065	0.3635030558377792
mom_octamer_mbuild.pdb_3	     6	12.01078	0.0000000000000000	A    	0.3480646886945065	0.3635030558377792
mom_octamer_mbuild.pdb_4	     6	12.01078	0.0000000000000000	A    	0.3480646886945065	0.3635030558377792
mom_octamer_mbuild.pdb_5	     6	12.01078	0.0000000000000000	A    	0.3480646886945065	0.3635030558377792
mom_octamer_mbuild.pdb_6	     6	12.01078	0.0000000000000000	A    	0.3480646886945065	0.3635030558377792
```

The `openff.toolkit.topology.molecule.Molecule` I feed to the Interchange object has the correct atom names, however, they aren't reflected in the final topology. Even the topology in the Interchange object has the right atom names.

```
$ interchange.topology.atom(1)
Atom(name=O2, atomic number=8)

```

I don't know if this is a bug or just that somewhere atom names are undefined and Interchange is defaulting to the filename convention shown here. Is this a specific case where these atom type names are used or is this default behavior? I ran into this issue with version 0.3.7.

**Comment by mattwthompson** [comment]
At first glance I think the issue here is atom names vs _atom type_ names. Atom types don't exist in OpenFF objects so they (and their names) are generated on the fly somewhere in the Interchange machinery. Atom names, however, do exist in OpenFF infrastructure and should be captured in the GROMACS files, I think in the `[ atoms ]` directive. It's easy to imagine Interchange should have better logic for creating these atom type names, however. I'm not sure if the periods or underscores are causing issues, but something like that could.

I can look at this more closely tomorrow, but my guess is that I decided atom type names should be generated a la `f"{molecule.name}_{i}` where `i` is some iterator while unique atoms or non-bonded interactions in the molecule are enumerated. My guess is that you have a molecule named `mom_octamer_mbuild.pdb` (the file name-looking thin having snuck into its `Molecule.name` attribute).

Could you share a minimal script that reproduces this? And, for context, is this providing issues at the level of engine crashes or user convenience (i.e. post-processing/visualization is more cumbersome)?

**Comment by tlfobe** [comment]
Hey Matt!

Thanks for taking a look at this. Here's a tar file with a minimal example of this. I'm assigning parameters to a terphenyl molecule here. I've included an SDF file with charges so that you don't have to generate them on your computer. I also included an `environment.yml` file to build the environment I'm working in and the final output files.

Currently, I'm not having any issues running gromacs simulations with these files, however, I was just confused because when I was using this workflow ~4-5 months ago atom type names were preserved and just mirrored atom names. 

[off_atomtyping_issue.tar.gz](https://github.com/openforcefield/openff-interchange/files/12148181/off_atomtyping_issue.tar.gz)

Thanks!
